CubeSQL_SDK/cubesql.c

```diff
@@ -936,185 +937,238 @@ void csql_dbfree (csqldb *db) {
+		sock_list[sock_index] = -1;
+		
+		// display protocol specific formatted address
+		/*
+		char szHost[256], szPort[16];
+		getnameinfo(addr->ai_addr, addr->ai_addrlen, szHost, sizeof(szHost), szPort, sizeof(szPort), NI_NUMERICHOST | NI_NUMERICSERV);
+		printf("getnameinfo(): host=%s, port=%s, family=%d\n", szHost, szPort, addr->ai_family);
+		*/
+		sock_current = socket(addr->ai_family, addr->ai_socktype, addr->ai_protocol);
+		if (sock_current < 0) continue;
+		
+		// set socket options
+		int len = 1;
+		bsd_setsockopt(sock_current, SOL_SOCKET, SO_KEEPALIVE, (const char *) &len, sizeof(len));
+		len = 1;
+		bsd_setsockopt(sock_current, IPPROTO_TCP, TCP_NODELAY, (const char *) &len, sizeof(len));
+		#ifdef SO_NOSIGPIPE
+		len = 1;
+		bsd_setsockopt(sock_current, SOL_SOCKET, SO_NOSIGPIPE, (const char *) &len, sizeof(len));
+		#endif
+		
+		// by default, an IPv6 socket created on Windows Vista and later only operates over the IPv6 protocol
+		// in order to make an IPv6 socket into a dual-stack socket, the setsockopt function must be called
+		if (addr->ai_family == AF_INET6) {
+			#ifdef WIN32
+			DWORD ipv6only = 0;
+			#else
+			int   ipv6only = 0;
+			#endif
+			bsd_setsockopt(sock_current, IPPROTO_IPV6, IPV6_V6ONLY, (void *)&ipv6only, sizeof(ipv6only));
+		}
+		
+		// turn on non-blocking
+		unsigned long ioctl_blocking = 1;    /* ~0; //TRUE; */
+		ioctl(sock_current, FIONBIO, &ioctl_blocking);
+		
+		// initiate non-blocking connect ignoring return code
+		rc = connect(sock_current, addr->ai_addr, addr->ai_addrlen);
+		if (rc < 0) {
+			if (errno != 0 && errno != EINTR && errno != EAGAIN & errno != EINPROGRESS)
+			{
+				// obvious error (e.g. network not reachable) - don't use this socket
+				lastConnectionErrorMessage = strerror(errno);
+				/*
+				printf("connect error: %s (%i)\n", strerror(errno), errno);
+				*/
+				closesocket(sock_current);
+				continue;
+			}
+		}
+		
+		// add sock_current to internal list of trying to connect sockets
+		sock_list[sock_index] = sock_current;
+	}
+	
+	// free not more needed memory
+	freeaddrinfo(addr_list);
 	
 	// calculate the connection timeout and reset timers
 	int connect_timeout = (db->timeout > 0) ? db->timeout : CUBESQL_DEFAULT_TIMEOUT;
 	time_t start = time(NULL);
-    time_t now = start;
-    rc = 0;
-    
-    int socket_err = 0;
-    int sockfd = 0;
-    fd_set write_fds;
-    fd_set except_fds;
-    struct timeval tv;
-    
-	while (rc == 0 && ((now - start) < connect_timeout)) {
+	time_t now = start;
+	rc = 0;
+	
+	int sockfd = 0;
+	fd_set write_fds;
+	fd_set except_fds;
+	struct timeval tv;
+
+	while ((now - start) < connect_timeout) {
 		FD_ZERO(&write_fds);
-        FD_ZERO(&except_fds);
-        
-        int nfds = 0;
-        for (int i=0; i<MAX_SOCK_LIST; ++i) {
-            if (sock_list[i]) {
-                FD_SET(sock_list[i], &write_fds);
-                FD_SET(sock_list[i], &except_fds);
-                if (nfds < sock_list[i]) nfds = sock_list[i];
```

Comment:
Main Issue: `nfds` is only assigned to the last socket in the loop (there may be previous sockets in the loop also trying to connect)